### PR TITLE
#454 - Verify the `at_hash` presence in the ID_token for 'id_token token' (Implicit) and 'code id_token token' (hybrid) flow.

### PR DESCRIPTION
**Pull request by duttarnab**
#454 - Verify the `at_hash` presence in the ID_token for 'id_token token' (Implicit) and 'code id_token token' (hybrid) flow.

https://github.com/GluuFederation/oxd/issues/454

### CODE REVIEW

File: oxd-server/src/main/java/org/gluu/oxd/server/op/CheckIdTokenOperation.java [hunk 1]
```diff
@@ -61,4 +69,15 @@ public IOpResponse execute(CheckIdTokenParams params) {
         throw HttpException.internalError();
     }
 
+    public static boolean atHashCheckRequired(List<String> responseTypes) {
```
Comment (yuriyz) [inline oxd-server/src/main/java/org/gluu/oxd/server/op/CheckIdTokenOperation.java:72]: `responseTypes.stream().anyMatch(s -> ResponseType.fromString(s, "").contains(TOKEN))`

File: oxd-server/src/main/java/org/gluu/oxd/server/op/Validator.java [hunk 2]
```diff
@@ -132,14 +132,24 @@ public Validator build() {
...
+    // `at_hash` in ID_TOKEN is mandatory for response_type='id_token token' for Implicit Flow or response_type='code id_token token' for Hybrid Flow.
+    public void validateAccessToken(String accessToken, boolean atHashRequired) {
         if (!Strings.isNullOrEmpty(accessToken)) {
             String atHash = idToken.getClaims().getClaimAsString("at_hash");
             if (Strings.isNullOrEmpty(atHash)) {
-                LOG.warn("Skip access_token validation because corresponding id_token does not have at_hash claim. access_token: " + accessToken + ", id_token: " + idToken);
-                return;
+                if(atHashRequired) {
+                    LOG.error("`at_hash` is missing in `ID_TOKEN`.");
+                    throw new HttpException(ErrorResponseCode.AT_HASH_NOT_FOUND);
```
Comment (yuriyz) [inline oxd-server/src/main/java/org/gluu/oxd/server/op/Validator.java:145]: Please add configuration properties `id_token_validation_at_hash_required: true`, ``id_token_validation_c_hash_required: true` and so on. Default value should be `true`.

File: oxd-server/src/main/java/org/gluu/oxd/server/op/CheckIdTokenOperation.java (hunk 1 above)
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/op/CheckIdTokenOperation.java:72]: Done

File: oxd-server/src/main/java/org/gluu/oxd/server/op/Validator.java (hunk 2 above)
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/op/Validator.java]: Done

**Pull request merged**